File: python/mirage/core/nextcloud/search/response.py
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from http import HTTPStatus
from xml.etree import ElementTree

from mirage.core.nextcloud.search.constants import (CONTENT_LENGTH,
                                                    DISPLAY_NAME,
                                                    LAST_MODIFIED,
                                                    RESOURCE_TYPE, SIZE)
from mirage.core.nextcloud.search.target import relative_path
from mirage.core.nextcloud.search.types import (Property, SearchEntry,
                                                SearchTarget, XmlElement)
from mirage.core.nextcloud.search.xml import dav
from mirage.types import FindType


def successful_status(status: str) -> bool:
    parts = status.split()
    return len(parts) >= 2 and parts[1] == str(HTTPStatus.OK.value)


def successful_properties(response: XmlElement) -> list[XmlElement]:
    properties: list[XmlElement] = []
    for propstat in response.findall(dav("propstat")):
        status = propstat.findtext(dav("status"), "")
        prop = propstat.find(dav("prop"))
        if successful_status(status) and prop is not None:
            properties.append(prop)
    if not properties:
        raise ValueError(
            "Nextcloud Files Search result has no successful properties")
    return properties


def find_text(properties: list[XmlElement], field: Property) -> str | None:
    for prop in properties:
        value = prop.findtext(field.tag)
        if value is not None:
            return value
    return None


def has_collection(properties: list[XmlElement]) -> bool:
    for prop in properties:
        resource_type = prop.find(RESOURCE_TYPE.tag)
        if (resource_type is not None
                and resource_type.find(dav("collection")) is not None):
            return True
    return False


def modified_timestamp(value: str | None) -> float | None:
    if value is None:
        return None
    try:
        modified = parsedate_to_datetime(value)
    except (TypeError, ValueError):
        try:
            modified = datetime.fromisoformat(value.replace("Z", "+00:00"))
        except ValueError as exc:
            raise ValueError(
                f"invalid Nextcloud Files Search timestamp: {value}") from exc
    if modified.tzinfo is None:
        modified = modified.replace(tzinfo=timezone.utc)
    return modified.timestamp()


def entry_size(properties: list[XmlElement]) -> int | None:
    value = find_text(properties, SIZE)
    if value is None:
        value = find_text(properties, CONTENT_LENGTH)
    if value is None:
        return None
    try:
        return int(value)
    except ValueError as exc:
        raise ValueError(
            f"invalid Nextcloud Files Search size: {value}") from exc


def parse_response(response: XmlElement, target: SearchTarget) -> SearchEntry:
    href = response.findtext(dav("href"))
    if href is None:
        raise ValueError("Nextcloud Files Search result is missing href")
    properties = successful_properties(response)
    key = relative_path(href, target)
    name = (find_text(properties, DISPLAY_NAME)
            or key.rstrip("/").rsplit("/", 1)[-1])
    return SearchEntry(
        key=key,
        name=name,
        kind=(FindType.DIRECTORY
              if has_collection(properties) else FindType.FILE),
        size=entry_size(properties),
        modified=modified_timestamp(find_text(properties, LAST_MODIFIED)),
    )


def parse_page(content: bytes, target: SearchTarget) -> list[SearchEntry]:
    root = ElementTree.fromstring(content)
    if root.tag != dav("multistatus"):
        raise ValueError("invalid Nextcloud Files Search response")
    return [
        parse_response(response, target)
        for response in root.findall(dav("response"))
    ]

File: typescript/packages/node/src/core/nextcloud/search/response.ts
import { XMLParser } from 'fast-xml-parser'
import { SyntaxValidator } from 'fast-xml-validator'
import { rstripSlash } from '@struktoai/mirage-core'
import { CONTENT_LENGTH, DISPLAY_NAME, LAST_MODIFIED, RESOURCE_TYPE, SIZE } from './constants.ts'
import { relativePath } from './target.ts'
import type { Property, SearchEntry, SearchTarget, XmlRecord } from './types.ts'

const parser = new XMLParser({
  ignoreAttributes: false,
  removeNSPrefix: true,
  parseTagValue: false,
  trimValues: false,
})

function isRecord(value: unknown): value is XmlRecord {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}

function records(value: unknown): XmlRecord[] {
  if (Array.isArray(value)) return value.filter(isRecord)
  return isRecord(value) ? [value] : []
}

function stringValue(value: unknown): string | null {
  if (typeof value === 'string' || typeof value === 'number') return String(value)
  return null
}

function successfulProperties(response: XmlRecord): XmlRecord[] {
  const properties: XmlRecord[] = []
  for (const propstat of records(response.propstat)) {
    const status = stringValue(propstat.status) ?? ''
    const fields = status.split(/\s+/)
    if (fields[1] === '200' && isRecord(propstat.prop)) properties.push(propstat.prop)
  }
  if (properties.length === 0) {
    throw new Error('Nextcloud Files Search result has no successful properties')
  }
  return properties
}

function findText(properties: XmlRecord[], field: Property): string | null {
  for (const propertySet of properties) {
    const value = stringValue(propertySet[field.name])
    if (value !== null) return value
  }
  return null
}

function hasCollection(properties: XmlRecord[]): boolean {
  return properties.some((propertySet) => {
    const resourceType = propertySet[RESOURCE_TYPE.name]
    return isRecord(resourceType) && 'collection' in resourceType
  })
}

function modifiedTimestamp(value: string | null): number | null {
  if (value === null) return null
  const milliseconds = Date.parse(value)
  if (Number.isNaN(milliseconds)) {
    throw new Error(`invalid Nextcloud Files Search timestamp: ${value}`)
  }
  return milliseconds / 1000
}

function entrySize(properties: XmlRecord[]): number | null {
  const value = findText(properties, SIZE) ?? findText(properties, CONTENT_LENGTH)
  if (value === null) return null
  const size = Number(value)
  if (!Number.isInteger(size)) throw new Error(`invalid Nextcloud Files Search size: ${value}`)
  return size
}

function parseResponse(response: XmlRecord, target: SearchTarget): SearchEntry {
  const href = stringValue(response.href)
  if (href === null) throw new Error('Nextcloud Files Search result is missing href')
  const properties = successfulProperties(response)
  const key = relativePath(href, target)
  return {
    key,
    name: findText(properties, DISPLAY_NAME) ?? rstripSlash(key).split('/').pop() ?? '',
    kind: hasCollection(properties) ? 'd' : 'f',
    size: entrySize(properties),
    modified: modifiedTimestamp(findText(properties, LAST_MODIFIED)),
  }
}

function validateXml(content: string): void {
  try {
    const validation = SyntaxValidator.validate(content)
    if (validation !== true) throw new Error(validation.err.msg)
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error)
    throw new Error(`invalid Nextcloud Files Search XML: ${message}`, { cause: error })
  }
}

export function parsePage(content: string, target: SearchTarget): SearchEntry[] {
  validateXml(content)
  const parsed = parser.parse(content) as unknown
  if (!isRecord(parsed) || !('multistatus' in parsed)) {
    throw new Error('invalid Nextcloud Files Search response')
  }
  const multistatus = parsed.multistatus
  if (multistatus === '') return []
  if (!isRecord(multistatus)) throw new Error('invalid Nextcloud Files Search response')
  return records(multistatus.response).map((response) => parseResponse(response, target))
}
